internal/codeintel/bundles/persistence/cache/reader_cache.go

```diff
@@ -0,0 +1,228 @@
+	//
+	// 🔐 - the cache mutex must be held when reading or modifying this value.
+	expiry time.Time
+
+	// draining marks the intent to close this entry. No additional readers
+	// of this entry should be allowed after this value is set to true.
+	//
+	// 🔐 - the cache mutex must be held when reading or modifying this value.
+	draining bool
+}
+
+// ReaderOpener initializes a new reader for the given key.
+type ReaderOpener func(key string) (persistence.Reader, error)
+
+// Handler performs an operation on a reader. The invocation of this function is a critical section that
+// locks the given reader argument so that it is not closed while in use.
+type Handler func(reader persistence.Reader) error
+
+// ErrReaderInitializationDeadlineExceeded occurs when a new reader takes too long to initialize.
+var ErrReaderInitializationDeadlineExceeded = errors.New("reader initialization deadline exceeded")
+
+// NewReaderCache initializes a new reader cache with the given TTL and reader opener.
+func NewReaderCache(ttl time.Duration, opener ReaderOpener) *Cache {
+	return newReaderCache(ttl, time.NewTicker(ttl).C, opener)
+}
+
+// newReaderCache initializes a new reader cache with the given ticker and reader opener.
+func newReaderCache(ttl time.Duration, ch <-chan time.Time, opener ReaderOpener) *Cache {
+	cache := &Cache{
+		ttl:     ttl,
+		opener:  opener,
+		entries: map[string]*cacheEntry{},
+	}
+
+	go func() {
+		for now := range ch {
+			cache.evict(now.UTC())
+		}
+	}()
+
+	return cache
+}
+
+// WithReader calls the given function with a reader argument. If the reader has not yet initialized and
+// does not do so before context deadline, an error is returned. The reader initialization is not canceled
+// due to a context deadline here and will continue to run in the background until completion.
+func (c *Cache) WithReader(ctx context.Context, key string, fn Handler) error {
+	if entry, ok := c.getOrCreateActiveEntry(ctx, key); ok {
+		defer func() {
+			// Decrement refcount
+			if atomic.AddUint32(&entry.refCount, ^uint32(0)) == 0 {
+				// Wake routines waiting for this value to go to zero
+				entry.refCountChanged.Broadcast()
+
+				// No more readers, update expiry
+				entry.expiry = time.Now().UTC().Add(c.ttl)
+			}
+		}()
+
+		select {
+		case <-entry.init:
+			if entry.err != nil {
+				return entry.err
+			}
+			return fn(entry.reader)
+
+		case <-ctx.Done():
+		}
+	}
+
+	return ErrReaderInitializationDeadlineExceeded
+}
+
+// getOrCreateActiveEntry gets or creates a cache entry for the given key. If there exists an entry for
+// the key, but that entry is marked as draining, we wait until that entry has completely drained and then
+// try again.
+func (c *Cache) getOrCreateActiveEntry(ctx context.Context, key string) (*cacheEntry, bool) {
+loop:
+	for {
+		entry, ok := c.getOrCreateRawEntry(key)
+		if ok {
+			return entry, true
+		}
+
+		select {
+		case <-entry.drained:
+			continue loop
```

Comment:
nit: I don't think [`continue` in Go's select is valid](https://play.golang.org/p/dX-_x8_rSGE) so the following should work too:
```suggestion
	for {
		entry, ok := c.getOrCreateRawEntry(key)
		if ok {
			return entry, true
		}

		select {
		case <-entry.drained:
			continue
```

Reply:
Yeah, the current version works, just not strictly necessary 😛